pkg/kubelet/cm/memorymanager/memory_manager.go

```diff
@@ -0,0 +1,314 @@
+func NewManager(policyName string, machineInfo *cadvisorapi.MachineInfo, nodeAllocatableReservation v1.ResourceList, stateFileDirectory string, affinity topologymanager.Store) (Manager, error) {
+	// TODO: we should add new kubelet parameter, and to get reserved memory per NUMA node from it
+	// currently we use kube-reserved + system-reserved + eviction reserve for each NUMA node, that creates memory over-consumption
+	// and no reservation for huge pages
+	reserved := ReservedMemory{}
+	for _, node := range machineInfo.Topology {
+		memory := nodeAllocatableReservation[v1.ResourceMemory]
+		if memory.IsZero() {
+			break
+		}
+		value, succeeded := memory.AsInt64()
+		if !succeeded {
+			return nil, fmt.Errorf("failed to represent reserved memory as int64")
+		}
+
+		reserved[node.Id] = map[v1.ResourceName]uint64{
+			v1.ResourceMemory: uint64(value),
+		}
+	}
+
+	var policy Policy
+
+	switch policyType(policyName) {
+
+	case policyTypeNone:
+		policy = NewPolicyNone()
+
+	case policyTypeSingleNUMA:
+		policy = NewPolicySingleNUMA(machineInfo, reserved, affinity)
+
+	default:
+		return nil, fmt.Errorf("unknown policy: \"%s\"", policyName)
+	}
+
+	manager := &manager{
+		policy:                     policy,
+		nodeAllocatableReservation: nodeAllocatableReservation,
+		stateFileDirectory:         stateFileDirectory,
+	}
+	manager.sourcesReady = &sourcesReadyStub{}
+	return manager, nil
+}
+
+// Start starts the memory manager reconcile loop under the kubelet to keep state updated
+func (m *manager) Start(activePods ActivePodsFunc, sourcesReady config.SourcesReady, podStatusProvider status.PodStatusProvider, containerRuntime runtimeService, initialContainers containermap.ContainerMap) error {
+	klog.Infof("[memorymanager] starting with %s policy", m.policy.Name())
+	m.sourcesReady = sourcesReady
+	m.activePods = activePods
+	m.podStatusProvider = podStatusProvider
+	m.containerRuntime = containerRuntime
+	m.containerMap = initialContainers
+
+	stateImpl, err := state.NewCheckpointState(m.stateFileDirectory, memoryManagerStateFileName, m.policy.Name(), m.containerMap)
+	if err != nil {
+		klog.Errorf("[memorymanager] could not initialize checkpoint manager: %v, please drain node and remove policy state file", err)
+		return err
+	}
+	m.state = stateImpl
+
+	err = m.policy.Start(m.state)
+	if err != nil {
+		klog.Errorf("[memorymanager] policy start error: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+// AddContainer saves the value of requested memory for the guaranteed pod under the state and set memory affinity according to the topolgy manager
+func (m *manager) AddContainer(pod *v1.Pod, container *v1.Container, containerID string) error {
+	// Get memory blocks assigned to the container during Allocate()
+	blocks := m.state.GetMemoryBlocks(string(pod.UID), container.Name)
+
+	if len(blocks) > 0 {
+		// it does not possible that memory blocks under the same container will have different NUMA affinity,
```

Comment:
I didnt take a look this part detaily before.
I think we do the below things rather than take NUMAAffinity from the first block.

1) if we won't introduce AddContainer Interface on policy interface.
- merging all NUMAAffinities of blocks of a container
- then just setting it on container's cgroup
  (this logic will work for cross-NUMA affinity later)

2)  introduce AddContainer function on Policy Interface
- calling out policy.AddContainer(...) here
- checking merged NUMAAffinity has single NUMA affintiy in a single-numa policy
- if it has more than one NUMA nodes we should reject it.